File: 37.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reorder_by_parity(head):
    if not head or not head.next:
        return head

    even_dummy = ListNode(0)
    odd_dummy = ListNode(0)
    even_tail, odd_tail = even_dummy, odd_dummy

    curr = head
    while curr:
        if curr.val % 2 == 0:
            even_tail.next = curr
            even_tail = even_tail.next
        else:
            odd_tail.next = curr
            odd_tail = odd_tail.next
        curr = curr.next

    # Connect even list to odd list
    even_tail.next = odd_dummy.next
    odd_tail.next = None

    return even_dummy.next


# --- Helper to print the list ---
def print_list(head):
    vals = []
    while head:
        vals.append(str(head.val))
        head = head.next
    print(" -> ".join(vals))


# --- Example ---
head = ListNode(1)
head.next = ListNode(4)
head.next.next = ListNode(5)
head.next.next.next = ListNode(2)
head.next.next.next.next = ListNode(7)
head.next.next.next.next.next = ListNode(8)

res = reorder_by_parity(head)
print_list(res)

File: 37.js
class ListNode {
  constructor(val, next = null) {
    this.val = val;
    this.next = next;
  }
}

function reorderByParity(head) {
  if (!head || !head.next) return head;

  let evenDummy = new ListNode(0);
  let oddDummy = new ListNode(0);
  let evenTail = evenDummy, oddTail = oddDummy;

  let curr = head;
  while (curr) {
    if (curr.val % 2 === 0) {
      evenTail.next = curr;
      evenTail = evenTail.next;
    } else {
      oddTail.next = curr;
      oddTail = oddTail.next;
    }
    curr = curr.next;
  }

  // Combine lists
  evenTail.next = oddDummy.next;
  oddTail.next = null;

  return evenDummy.next;
}

// --- Helper to print list ---
function printList(head) {
  let arr = [];
  while (head) {
    arr.push(head.val);
    head = head.next;
  }
  console.log(arr.join(" -> "));
}

// --- Example ---
let head = new ListNode(1);
head.next = new ListNode(4);
head.next.next = new ListNode(5);
head.next.next.next = new ListNode(2);
head.next.next.next.next = new ListNode(7);
head.next.next.next.next.next = new ListNode(8);

let result = reorderByParity(head);
printList(result);
